### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by ruancomelli**
# Description

_Explain briefly about the purpose & context of this PR_

# Screenshots

_If possible add a screenshot and/or gif_

# Example | Usage

_Please share examples and how to reproduce_

---

# Related Issues | Pr

_`close`|`resolve` <ISSUE> (ex. `resolve #20`)_
_link issues to close automatically_

### CODE REVIEW

**Comment by ruancomelli** [comment]
This is an experiment review for experiment `mermaid_diagrams`
Run ID: `mermaid_diagrams/run_2024-09-20T10-59-17_v1-22-0-133-gbcf338da5-dirty`
The benchmark review for this pull request can be found at https://github.com/sourcery-ai-experiments/timba-api/pull/2.
This pull request was cloned from `https://github.com/AlphaDev87/timba-api/pull/49`. _(Note: the URL is not a link to avoid triggering a notification on the original pull request.)_

<details>
<summary>Experiment configuration</summary>

```yaml
review_config:
  # User configuration for the review
  # - benchmark - use the user config from the benchmark reviews
  # - <value> - use the value directly
  user_review_config:
    enable_ai_review: true
    enable_rule_comments: false

    enable_complexity_comments: false
    enable_security_comments: false
    enable_tests_comments: false
    enable_comment_suggestions: false
    enable_functionality_review: false

    enable_pull_request_summary: false
    enable_review_guide: true

    enable_approvals: false

  ai_review_config:
    # The model responses to use for the experiment
    # - benchmark - use the model responses from the benchmark reviews
    # - llm - call the language model to generate responses
    model_responses:
      comments_model: benchmark
      comment_area_model: benchmark
      comment_validation_model: benchmark
      comment_suggestion_model: benchmark
      complexity_model: benchmark
      functionality_model: benchmark
      security_model: benchmark
      tests_model: benchmark
      pull_request_summary_model: benchmark
      review_guide_model: llm
      overall_comments_model: benchmark

# The pull request dataset to run the experiment on
pull_request_dataset:
- https://github.com/sourcery-ai/core/pull/4607
- https://github.com/sourcery-ai/core/pull/4631
- https://github.com/sourcery-ai/core/pull/4647
    # CodeRabbit examples:
- https://github.com/2lambda123/-Orange-OpenSource-oorobot/pull/15
- https://github.com/2lambda123/galaxyproject-galaxy/pull/12
- https://github.com/a0v0/avtoolz/pull/79
- https://github.com/adityask98/Hotaru/pull/10
- https://github.com/agdas/vscode/pull/2
- https://github.com/agluszak/hirschgarten/pull/2
- https://github.com/alexsnow348/insightface/pull/46
- https://github.com/alikuxac/utilities/pull/10
- https://github.com/AlphaDev87/timba-api/pull/49
- https://github.com/AngeloTadeucci/Maple2/pull/239
- https://github.com/AngeloTadeucci/Maple2.File/pull/36
- https://github.com/AngeloTadeucci/Maple2/pull/233
  # Examples where CodeRabbit does not generate diagrams
- https://github.com/baptisteArno/typebot.io/pull/1778
- https://github.com/btipling/foundations/pull/33
- https://github.com/btipling/foundations/pull/31
- https://github.com/chintu-777/jaeger/pull/1
- https://github.com/coji/remix-docs-ja/pull/55
- https://github.com/DaveMBush/SmartNgRX/pull/622
- https://github.com/DaveMBush/SmartNgRX/pull/481
- https://github.com/dkittle/party-connections/pull/6
- https://github.com/Drajad-Kusuma-Adi/onstudy-backend/pull/6
- https://github.com/imaami/libcanth/pull/2

# Questions to ask to label the review comments
review_comment_labels: []
# - label: correct
#   question: Is this comment correct?

# Benchmark reviews generated by running
#   python -m scripts.experiment benchmark <experiment_name>
benchmark_reviews:
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4607
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/338
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4631
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/339
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4647
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/340
- dataset_pull_request: https://github.com/2lambda123/-Orange-OpenSource-oorobot/pull/15
  review_pull_request: https://github.com/sourcery-ai-experiments/-Orange-OpenSource-oorobot/pull/2
- dataset_pull_request: https://github.com/2lambda123/galaxyproject-galaxy/pull/12
  review_pull_request: https://github.com/sourcery-ai-experiments/galaxyproject-galaxy/pull/1
- dataset_pull_request: https://github.com/adityask98/Hotaru/pull/10
  review_pull_request: https://github.com/sourcery-ai-experiments/Hotaru/pull/2
- dataset_pull_request: https://github.com/agdas/vscode/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/vscode/pull/3
- dataset_pull_request: https://github.com/agluszak/hirschgarten/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/hirschgarten/pull/2
- dataset_pull_request: https://github.com/alikuxac/utilities/pull/10
  review_pull_request: https://github.com/sourcery-ai-experiments/utilities/pull/2
- dataset_pull_request: https://github.com/AlphaDev87/timba-api/pull/49
  review_pull_request: https://github.com/sourcery-ai-experiments/timba-api/pull/2
- dataset_pull_request: https://github.com/AngeloTadeucci/Maple2/pull/239
  review_pull_request: https://github.com/sourcery-ai-experiments/Maple2/pull/3
- dataset_pull_request: https://github.com/AngeloTadeucci/Maple2.File/pull/36
  review_pull_request: https://github.com/sourcery-ai-experiments/Maple2.File/pull/2
- dataset_pull_request: https://github.com/AngeloTadeucci/Maple2/pull/233
  review_pull_request: https://github.com/sourcery-ai-experiments/Maple2/pull/4
- dataset_pull_request: https://github.com/baptisteArno/typebot.io/pull/1778
  review_pull_request: https://github.com/sourcery-ai-experiments/typebot.io/pull/1
- dataset_pull_request: https://github.com/btipling/foundations/pull/33
  review_pull_request: https://github.com/sourcery-ai-experiments/foundations/pull/1
- dataset_pull_request: https://github.com/btipling/foundations/pull/31
  review_pull_request: https://github.com/sourcery-ai-experiments/foundations/pull/2
- dataset_pull_request: https://github.com/chintu-777/jaeger/pull/1
  review_pull_request: https://github.com/sourcery-ai-experiments/jaeger/pull/1
- dataset_pull_request: https://github.com/coji/remix-docs-ja/pull/55
  review_pull_request: https://github.com/sourcery-ai-experiments/remix-docs-ja/pull/1
- dataset_pull_request: https://github.com/DaveMBush/SmartNgRX/pull/622
  review_pull_request: https://github.com/sourcery-ai-experiments/SmartNgRX/pull/1
- dataset_pull_request: https://github.com/DaveMBush/SmartNgRX/pull/481
  review_pull_request: https://github.com/sourcery-ai-experiments/SmartNgRX/pull/2
- dataset_pull_request: https://github.com/dkittle/party-connections/pull/6
  review_pull_request: https://github.com/sourcery-ai-experiments/party-connections/pull/1
- dataset_pull_request: https://github.com/Drajad-Kusuma-Adi/onstudy-backend/pull/6
  review_pull_request: https://github.com/sourcery-ai-experiments/onstudy-backend/pull/1
- dataset_pull_request: https://github.com/imaami/libcanth/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/libcanth/pull/1

```

</details>

**Comment by SourceryAI** [comment]
## Reviewer's Guide by Sourcery

This pull request implements significant changes to the project structure, database schema, and functionality. Key changes include:

1. Refactoring of the transaction system, splitting it into separate deposit and payment components.
2. Introduction of new features such as password reset, analytics, and bot management.
3. Improvements to error handling and API responses.
4. Updates to the database schema, including new tables and modifications to existing ones.
5. Enhanced security measures and rate limiting.
6. Refactoring of various services and controllers for better organization and separation of concerns.

### File-Level Changes

| Change | Details | Files |
| ------ | ------- | ----- |
| Refactored transaction system | <ul><li>Split transactions into separate deposit and payment components</li><li>Created new DepositServices and PaymentServices classes</li><li>Updated related controllers and routes</li><li>Modified database schema for deposits and payments</li></ul> | `src/components/deposits/services.ts`<br/>`src/components/payments/services.ts`<br/>`src/components/deposits/controller.ts`<br/>`src/components/payments/controller.ts`<br/>`src/routes/transactions/deposits.router.ts`<br/>`src/routes/transactions/payments.router.ts` |
| Implemented password reset functionality | <ul><li>Added NewPasswordServices class for password reset logic</li><li>Created new routes for forgot password and reset password</li><li>Implemented email sending for password reset</li><li>Added new database table for password reset tokens</li></ul> | `src/components/new-password/services.ts`<br/>`src/components/new-password/controller.ts`<br/>`src/routes/auth.router.ts`<br/>`src/db/password-reset-token.ts`<br/>`prisma/migrations/20240508181836_initial_migration/migration.sql` |
| Added analytics feature | <ul><li>Created new AnalyticsDAO for database operations</li><li>Implemented AnalyticsController for handling analytics requests</li><li>Added new routes for analytics</li><li>Created new database table for analytics</li></ul> | `src/db/analytics.ts`<br/>`src/components/analytics/controller.ts`<br/>`src/routes/analytics.router.ts`<br/>`prisma/migrations/20240605135628_create_analytics_table/migration.sql` |
| Improved error handling and API responses | <ul><li>Created CustomError class for standardized error handling</li><li>Updated error middleware to use CustomError</li><li>Modified API responses to include status code and error details</li><li>Implemented specific error classes for different types of errors</li></ul> | `src/helpers/error/CustomError.ts`<br/>`src/middlewares/errorHandler.ts`<br/>`src/helpers/apiResponse.ts`<br/>`src/helpers/error/AgentApiError.ts`<br/>`src/helpers/error/AlquimiaApiError.ts` |
| Enhanced security measures | <ul><li>Implemented rate limiting for password reset and payment requests</li><li>Added validation for bank account numbers (CLABE)</li><li>Improved input validation across various endpoints</li><li>Updated token handling and expiration logic</li></ul> | `src/middlewares/rate-limiters/pass-restore-request.ts`<br/>`src/middlewares/rate-limiters/payment.ts`<br/>`src/components/bank-accounts/validators.ts`<br/>`src/services/jwt.service.ts` |
| Refactored and improved various services | <ul><li>Created ResourceService as a base class for common service operations</li><li>Updated PlayerServices, AgentServices, and other services to use new patterns</li><li>Improved typing and error handling in services</li><li>Implemented new services for Alquimia transfers and WhatsApp notifications</li></ul> | `src/services/resource.service.ts`<br/>`src/components/players/services.ts`<br/>`src/components/agent/services.ts`<br/>`src/services/alquimia-transfer.service.ts`<br/>`src/notification/whatsapp.ts` |

### Sequence Diagrams

#### Password Reset Flow
```mermaid
sequenceDiagram
    participant U as User
    participant A as API
    participant DB as Database
    participant E as Email Service
    U->>A: Request password reset
    A->>DB: Create reset token
    A->>E: Send reset email
    E-->>U: Receive reset email
    U->>A: Submit new password
    A->>DB: Verify token & update password
    A-->>U: Confirm password reset
```

#### Deposit Flow
```mermaid
sequenceDiagram
    participant U as User
    participant A as API
    participant D as DepositServices
    participant C as CasinoCoinsService
    participant DB as Database
    U->>A: Create deposit request
    A->>D: Process deposit
    D->>DB: Create deposit entry
    D->>C: Transfer coins
    C-->>D: Transfer result
    D->>DB: Update deposit status
    D-->>A: Deposit result
    A-->>U: Deposit confirmation
```

---

<details>
<summary>Tips</summary>

- Trigger a new Sourcery review by commenting `@sourcery-ai review` on the pull request.
- Continue your discussion with Sourcery by replying directly to review comments.
- You can change your review settings at any time by accessing your [dashboard](https://app.sourcery.ai):
    - Enable or disable the Sourcery-generated pull request summary or reviewer's guide;
    - Change the review language;
- You can always [contact us](mailto:support@sourcery.ai) if you have any questions or feedback.

</details>

File: src/db/players.ts [hunk 1]
```diff
@@ -1,26 +1,73 @@
...
+    try {
+      return prisma.player.count();
+    } catch (e) {
+      throw e;
+    } finally {
+      prisma.$disconnect();
+    }
+  }
+
+  static _getAll = async (
```
Comment (SourceryAI) [inline src/db/players.ts:30]: **suggestion (performance):** Optimize database operations for performance

With the introduction of more granular database operations, ensure that frequently called methods like _getAll are optimized for performance, especially when dealing with large datasets.

File: docker-compose.yml [hunk 2]
```diff
@@ -1,21 +1,20 @@
...
 services:
-  casinomex_db:
-    image: mysql
+  db:
+    image: mariadb
     restart: unless-stopped
     environment:
-      MYSQL_ROOT_PASSWORD: timba2024
-      MYSQL_DATABASE: timba-data-base
+      MARIADB_ROOT_PASSWORD: timba2024root
```
Comment (SourceryAI) [inline docker-compose.yml:7]: **🚨 suggestion (security):** Consider using environment variables for sensitive information

Storing passwords directly in the docker-compose.yml file can be a security risk, especially if the file is committed to version control. Consider using environment variables or a secrets management system for sensitive information.

**Review by SourceryAI — COMMENTED** [review]
Hey @ruancomelli - I've reviewed your changes and they look great!

<details>
<summary>Here's what I looked at during the review</summary>

- 🟡 **General issues**: 1 issue found
- 🟢 **Documentation**: all looks good
</details>

***

[LangSmith trace](https://smith.langchain.com/o/7d8518ad-6dd3-5c7b-96d2-757769296335/projects/p/88560225-5072-48bc-967c-dc54da53f265/r/ae420954-a135-49ea-abc3-08b2c3ddb122?poll=true)

<sub>
Help me be more useful! Please click 👍 or 👎 on each comment to tell me if it was helpful.
</sub>

**Pull request closed without merging**